tests/tx/scripts/test_sighash.py

```diff
@@ -0,0 +1,321 @@
+
+        # Bob signs his input using sighash_all to complete the tx
+        data_to_sign2 = atomic_swap_tx_clone.get_sighash_all()
+        assert self.manager2.wallet
+        public_bytes2, signature2 = self.manager2.wallet.get_input_aux_data(data_to_sign2, self.genesis_private_key)
+        htr_input.data = P2PKH.create_input_data(public_bytes2, signature2)
+
+        # The atomic swap tx is not valid and cannot be propagated
+        with pytest.raises(InvalidInputData) as e:
+            self.manager1.verification_service.verify(atomic_swap_tx_clone)
+
+        self.assertEqual(str(e.value), "Maximum number of outputs exceeded (4 > 3).")
+
+        with pytest.raises(InvalidNewTransaction):
+            self.manager1.propagate_tx(atomic_swap_tx_clone, fails_silently=False)
+
+    @patch('hathor.transaction.scripts.opcode.is_opcode_valid', lambda _: True)
+    def test_sighash_bitmask_input_not_selected(self) -> None:
+        # Create a new test token
+        token_creation_tx = create_tokens(self.manager1, self.address1_b58)
+        token_uid = token_creation_tx.tokens[0]
+        parents = self.manager1.get_new_tx_parents()
+
+        # Alice creates an input spending all created test tokens
+        tokens_input = TxInput(not_none(token_creation_tx.hash), 0, b'')
+
+        # Alice creates an input spending all genesis HTR
+        genesis_input = TxInput(not_none(self.genesis_block.hash), 0, b'')
+
+        # Alice creates an atomic swap tx
+        atomic_swap_tx = Transaction(
+            weight=1,
+            inputs=[tokens_input, genesis_input],
+            outputs=[],
+            parents=parents,
+            tokens=[token_uid],
+            storage=self.manager1.tx_storage,
+            timestamp=token_creation_tx.timestamp + 1
+        )
+        self.manager1.cpu_mining_service.resolve(atomic_swap_tx)
+
+        # Alice signs her token input using sighash bitmasks, instead of sighash_all.
+        sighash_bitmask = SighashBitmask(inputs=0b01, outputs=0b00)
+        data_to_sign1 = atomic_swap_tx.get_custom_sighash_data(sighash_bitmask)
+        assert self.manager1.wallet
+        public_bytes1, signature1 = self.manager1.wallet.get_input_aux_data(data_to_sign1, self.private_key1)
+        tokens_input.data = P2PKH.create_input_data(
+            public_key_bytes=public_bytes1,
+            signature=signature1,
+            sighash=sighash_bitmask,
+        )
+
+        # Alice signs her genesis input using the same sighash, so the genesis input is not selected in the bitmask.
+        public_bytes1, signature1 = self.manager1.wallet.get_input_aux_data(data_to_sign1, self.genesis_private_key)
+        genesis_input.data = P2PKH.create_input_data(
+            public_key_bytes=public_bytes1,
+            signature=signature1,
+            sighash=sighash_bitmask,
+        )
+
+        # The inputs are invalid, since one of them doesn't select itself.
+        with pytest.raises(InvalidInputData) as e:
```

Comment:
Shouldn't we create a specific exception for this case?

Reply:
Same as above, please see https://github.com/HathorNetwork/hathor-core/pull/912#discussion_r1470447371